src/Dependencies/PooledObjects/ArrayBuilder.cs

```diff
@@ -322,6 +322,13 @@ public ImmutableArray<U> ToDowncastedImmutable<U>()
             return tmp.ToImmutableAndFree();
         }
 
+        public ImmutableArray<U> ToDowncastedImmutableAndFree<U>() where U : T
+        {
+            var result = ToDowncastedImmutable<U>();
+            this.Free();
+            return result;
+        }
```

Comment:
i don't feel strongly about having this extension.  would be ok inlining it.  But it fits the pattern of how we use ArrayBuilders for the other ops, so i kept this in sync with those.